Add a strain dependent rheology model

naliboff opened the pull request:
Pull Request Checklist. Please read and check each box with an X. Delete any part not applicable. Ask on the [forum](https://community.geodynamics.org/c/aspect) if you need help with any step.

This PR moves code related to tracking of finite strain and strain softening behavior from the `ViscoPlastic` material model to a new model `StrainDependent` in the `Rheology` class. By and large, very little code had to be modified during the transfer. Looking through the existing code I see lots of room for improvement, but I think that should be done in follow-up pull requests.

All the relevant tests pass on my local machine with the current changes, but we'll see what the tester says.

### For all pull requests:

* [x] I have followed the [instructions for indenting my code](../CONTRIBUTING.md#making-aspect-better).

### For new features/models or changes of existing features:

* [X] I have tested my new feature locally to ensure it is correct.
* [ ] I have [created a testcase](http://www.math.clemson.edu/~heister/manual.pdf#sec%3Awriting_tests) for the new feature/benchmark in the [tests/](../tests/) directory.
* [X] I have added a changelog entry in the [doc/modules/changes](../doc/modules/changes) directory that will inform other users of my change.

gassmoeller commented on include/aspect/material_model/rheology/strain_dependent.h:
```diff
@@ -0,0 +1,221 @@
+           * A function that fills the reaction terms for the finite strain invariant(s) in
+           * MaterialModelOutputs object that is handed over.
+           */
+          void fill_reaction_outputs (const MaterialModel::MaterialModelInputs<dim> &in,
```
Perspectively, I would like to move this somewhere else (maybe make another sub-structure like `reactions`), but this would require to pass the necessary information around (like what type of strain should be tracked). So for now lets keep it together with the rest of the finite strain stuff like you did here.

gassmoeller commented on include/aspect/material_model/rheology/strain_dependent.h:
```diff
@@ -0,0 +1,221 @@
+           */
+          bool use_viscous_strain_weakening;
+
+          bool use_finite_strain_tensor;
```
Preexisting, but could you add documentation here?

gassmoeller commented on source/material_model/visco_plastic.cc:
```diff
@@ -493,84 +383,17 @@ namespace aspect
-        {
-          if (weakening_mechanism == plastic_weakening_with_plastic_strain_only || weakening_mechanism == plastic_weakening_with_plastic_strain_and_viscous_weakening_with_viscous_strain)
-            composition_mask.set(this->introspection().compositional_index_for_name("plastic_strain"),false);
+      // Return composition mask for strain fields
```
The comments in this function are not really clear to me. I liked the previous version: `Store which components to exclude during volume fraction computation.`, because it explains what this composition mask *is used for*, not what it is. Please just revert to the old comment.

naliboff commented on include/aspect/material_model/rheology/strain_dependent.h:
```diff
@@ -0,0 +1,221 @@
+           */
+          std::array<double, 3>
+          compute_weakened_yield_parameters(const unsigned int j,
+                                            const std::vector<double> cohesions,
```
This was one of the items on my cleanup list :+1:. Would you like this done in a follow-up PR or a separate commit here? No preference from my end.

naliboff commented:
@gassmoeller - Thanks for the review! I addressed all of the comments and modified the strain weakening functions such that the cohesion and friction do not need to be passed back and forth (third commit).

naliboff commented:
@jdannberg - Do you think this is a reasonable time  to remove backwards compatibility for the old strain weakening naming schemes in a follow-up PR?

gassmoeller commented on include/aspect/material_model/rheology/strain_dependent.h:
```diff
@@ -0,0 +1,226 @@
+          bool use_viscous_strain_weakening;
+
+          /**
+           * Where to track the full finite strain tensor and use an
```
```suggestion
           * Whether to track the full finite strain tensor and use
```

jdannberg commented:
> Do you think this is a reasonable time to remove backwards compatibility for the old strain weakening naming schemes in a follow-up PR?

Yes! I actually looked through the pull request on Friday, but then didn't have anything else to say in addition to what @gassmoeller already commented on, except for removing all of these old bools, which makes sense as a separate pull request.

naliboff commented:
@jdannberg - Great, I'll remove all those extra parameters and associated code next!

gassmoeller commented:
Ready to go.
